File: DSA/GreedyAlgo/Code/CoinChange.cs
// Coin Change Problem in C#

using System;
using System.Collections.Generic;
using System.Linq;

class CoinChange {
    public static void CoinChangeProblem(int[] coins, int amount) {
        Array.Sort(coins);
        Array.Reverse(coins);
        
        Console.WriteLine($"Coin Change for amount {amount}:");
        int coinsUsed = 0;
        
        foreach (int coin in coins) {
            while (amount >= coin) {
                Console.WriteLine($"Take coin of value {coin}");
                amount -= coin;
                coinsUsed++;
            }
        }
        
        Console.WriteLine($"\nTotal Coins Used: {coinsUsed}");
        if (amount > 0) {
            Console.WriteLine($"Warning: Cannot make exact change (Remaining: {amount})");
        } else {
            Console.WriteLine("Amount Made: Complete");
        }
    }
    
    static void Main() {
        Console.WriteLine("=== Coin Change Problem (C#) ===\n");
        
        int[] coins = {1, 5, 10, 25};
        int amount = 41;
        
        Console.Write("Available Coins: ");
        foreach (int coin in coins) {
            Console.Write(coin + " ");
        }
        Console.WriteLine($"\nAmount to Make: {amount}\n");
        
        CoinChangeProblem(coins, amount);
        
        Console.WriteLine("\n=== Greedy Strategy ===");
        Console.WriteLine("1. Sort coins in descending order");
        Console.WriteLine("2. For each coin (largest to smallest):");
        Console.WriteLine("   - Use as many as possible");
        Console.WriteLine("   - Subtract from amount");
        Console.WriteLine("3. Continue until amount becomes 0\n");
        
        Console.WriteLine("=== Complexity ===");
        Console.WriteLine("Time Complexity:  O(n log n) for sorting + O(amount)");
        Console.WriteLine("Space Complexity: O(1)");
        Console.WriteLine("Note: Greedy works for standard coin systems");
    }
}

File: DSA/GreedyAlgo/Code/CoinChange.py
# Coin Change Problem in Python

def coin_change(coins, amount):
    coins.sort(reverse=True)
    
    print(f"Coin Change for amount {amount}:")
    coins_used = 0
    
    for coin in coins:
        while amount >= coin:
            print(f"Take coin of value {coin}")
            amount -= coin
            coins_used += 1
    
    print(f"\nTotal Coins Used: {coins_used}")
    if amount > 0:
        print(f"Warning: Cannot make exact change (Remaining: {amount})")
    else:
        print("Amount Made: Complete")

# Test
print("=== Coin Change Problem (Python) ===\n")

coins = [1, 5, 10, 25]
amount = 41

print("Available Coins:", coins)
print(f"Amount to Make: {amount}\n")

coin_change(coins, amount)

print("\n=== Greedy Strategy ===")
print("1. Sort coins in descending order")
print("2. For each coin (largest to smallest):")
print("   - Use as many as possible")
print("   - Subtract from amount")
print("3. Continue until amount becomes 0\n")

print("=== Complexity ===")
print("Time Complexity:  O(n log n) for sorting + O(amount)")
print("Space Complexity: O(1)")
print("Note: Greedy works for standard coin systems")
